app/services/plugin.service.ts

```diff
@@ -1,236 +1,61 @@
-                                        } else {
-                                            return context.sinap.__result;
-                                        }
-                                    });
-                            },
-                            compilationMessages: context.sinap.__program.compilationMessages
-                        };
-                    }
-                });
-        })
     }
 
-    public getPlugin(kind: string): Promise<ConcretePlugin> {
-        const loadedPlugin = this.plugins.get(kind);
-        if (loadedPlugin) {
-            return loadedPlugin;
+    public getPlugin(kind: string) {
+        let plugin = this.plugins.get(kind);
+        if (plugin) {
+            return plugin;
+        }
+        const fileName = this.pluginKinds.get(kind);
+        if (!fileName) {
+            throw "Plugin not installed";
         }
-        const plugin = this.makePlugin(kind);
+        plugin = loadPlugin(fileName);
         this.plugins.set(kind, plugin);
         return plugin;
     }
 
-    public makePlugin(kind: string): Promise<ConcretePlugin> {
-        const val = this.pluginKinds.get(kind);
-
-        if (!val) {
-            return Promise.reject(`Unsupported plugin kind: ${kind}`);
-        }
-
-        const {definitions, interpreter} = val;
-
-        const definitionsData = this.fileService.fileByName(definitions)
-            .then((defFile: File) => defFile.readData())
-            .then((s) => {
-                return this.loadPluginTypeDefinitions(s);
-            });
-
-        const script = this.fileService.fileByName(interpreter)
-            .then((file: File) => file.readData())
-            .then((code) => this.sandboxService.compileScript(code));
-        return Promise.all([definitionsData, script])
-            .then(([def, scr]) => new ConcretePlugin(def, scr));
+    public getProgram(plugin: Plugin, m: CoreModel): Promise<Program> {
+        return this.getProgramContext(plugin).then(
+            context => new context.global['plugin-stub'].Program(m.serialize()));
     }
 
-    public loadPluginTypeDefinitions(src: string): Definitions {
-        const scope = parseScope(src);
-        scope.validate();
-
-        const nodes: [string, ClassMetaType][] = [];
-        const edges: [string, ClassMetaType][] = [];
-        const graphs: [string, ClassMetaType][] = [];
-
-        for (const [name, value] of scope.definitions.entries()) {
-            if (value.subtype(Type.Node)) {
-                nodes.push([name, value as ClassMetaType]);
-            }
-            if (value.subtype(Type.Edge)) {
-                edges.push([name, value as ClassMetaType]);
-            }
-            if (value.subtype(Type.Graph)) {
-                graphs.push([name, value as ClassMetaType]);
-            }
+    private getProgramContext(plugin: Plugin) {
+        let contextPromise = this.programs.get(plugin);
+        if (contextPromise === undefined) {
```

Comment:
Change this to just be if(contextPromise) to avoid weird issues with null vs. undefined.